File: flutter/packages/aval_format/test/graph_adapter_test.dart
// Dart port of packages/format/test/graph-adapter.test.ts.
//
// `GraphStateDefinition`/`GraphEdgeDefinition`/etc. in aval_graph do not
// override `==`, so this port asserts field-by-field rather than one deep
// `toEqual`, and drops the `Object.isFrozen`-style checks (no Dart
// equivalent; aval_graph's `List.unmodifiable` copies play the same
// "cannot be caller-mutated" role structurally).
import 'package:aval_format/src/errors.dart';
import 'package:aval_format/src/graph_adapter.dart';
import 'package:aval_format/src/manifest_schema.dart';
import 'package:aval_graph/aval_graph.dart';
import 'package:test/test.dart';

import 'manifest_fixture.dart';

void main() {
  group('adaptManifestToMotionGraph', () {
    test('maps the complete manifest graph to the hand-written M3 golden', () {
      final graph = adaptManifestToMotionGraph(validateCompiledManifestV01(validManifest()));
      final definition = graph.definition;

      expect(definition.initialState, 'a-a');
      expect(definition.states.map((s) => s.id).toList(), ['a-a', 'a-b', 'a-c']);

      final stateA = definition.states[0];
      expect(stateA.body.unitId, 'body-a');
      expect(stateA.body.kind, GraphBodyKind.loop);
      expect(stateA.body.frameCount, 4);
      expect(stateA.body.ports[0].id, 'default');
      expect(stateA.body.ports[0].portalFrames, [0, 2]);
      expect(stateA.initialUnit?.unitId, 'intro-a');
      expect(stateA.initialUnit?.frameCount, 2);

      final stateB = definition.states[1];
      expect(stateB.body.kind, GraphBodyKind.finite);
      expect(stateB.body.frameCount, 3);
      expect(stateB.body.ports[0].portalFrames, [2]);
      expect(stateB.initialUnit, isNull);

      final stateC = definition.states[2];
      expect(stateC.body.kind, GraphBodyKind.held);
      expect(stateC.body.frameCount, 1);
      expect(stateC.body.ports[0].portalFrames, [0]);

      expect(definition.edges.map((e) => e.id).toList(), [
        'edge-ab',
        'edge-ac',
        'edge-ba',
        'edge-bc',
        'edge-cb',
      ]);

      final edgeAb = definition.edges[0];
      expect(edgeAb.from, 'a-a');
      expect(edgeAb.to, 'a-b');
      expect((edgeAb.trigger as GraphEdgeTriggerEvent).name, 'go-b');
      final startAb = edgeAb.start as GraphStartPolicyPortal;
      expect(startAb.sourcePort, 'default');
      expect(startAb.targetPort, 'default');
      expect(startAb.maxWaitFrames, 1);
      final transitionAb = edgeAb.transition as GraphTransitionLocked;
      expect(transitionAb.unitId, 'bridge-ab');
      expect(transitionAb.frameCount, 2);
      expect(edgeAb.continuity, GraphContinuity.exactAuthored);

      final edgeAc = definition.edges[1];
      expect(edgeAc.start, isA<GraphStartPolicyCut>());
      expect(edgeAc.continuity, GraphContinuity.cut);
      expect(edgeAc.transition, isNull);

      final edgeBa = definition.edges[2];
      expect(edgeBa.trigger, isA<GraphEdgeTriggerCompletion>());
      expect(edgeBa.start, isA<GraphStartPolicyFinish>());

      final edgeBc = definition.edges[3];
      final transitionBc = edgeBc.transition as GraphTransitionReversible;
      expect(transitionBc.unitId, 'rev-bc');
      expect(transitionBc.frameCount, 6);
      expect(transitionBc.direction, TransitionDirection.forward);
      expect(transitionBc.reverseOf, isNull);
      expect(edgeBc.continuity, GraphContinuity.exactAuthored);

      final edgeCb = definition.edges[4];
      final transitionCb = edgeCb.transition as GraphTransitionReversible;
      expect(transitionCb.direction, TransitionDirection.reverse);
      expect(transitionCb.reverseOf, 'edge-bc');
      expect(edgeCb.continuity, GraphContinuity.exactReverse);
    });

    test('returns a graph detached from the manifest', () {
      final manifest = validateCompiledManifestV01(validManifest());
      final graph = adaptManifestToMotionGraph(manifest);

      expect(graph.definition.states, isNot(same(manifest.states)));
    });

    test('wraps M3 geometry and ambiguity failures as GRAPH_INVALID', () {
      final manifest = validManifest();
      final edges = (manifest['edges'] as List).cast<Map<String, Object?>>();
      final edge0 = Map<String, Object?>.from(edges[0]);
      final start = Map<String, Object?>.from(edge0['start'] as Map);
      start['maxWaitFrames'] = 0;
      edge0['start'] = start;
      final newEdges = [edge0, ...edges.skip(1)];
      final mutated = Map<String, Object?>.from(manifest)..['edges'] = newEdges;

      final schemaValid = validateCompiledManifestV01(mutated);
      expect(
        () => adaptManifestToMotionGraph(schemaValid),
        throwsA(predicate((e) => e is FormatError && e.code == FormatErrorCode.graphInvalid)),
      );
    });
  });
}

File: packages/format/test/graph-adapter.test.ts
import { describe, expect, it } from "vitest";

import { FormatError } from "../src/errors.js";
import { adaptManifestToMotionGraph } from "../src/graph-adapter.js";
import { validateCompiledManifest } from "../src/manifest-schema.js";
import type { CompiledManifest } from "../src/model.js";
import { validManifest } from "./manifest-fixture.js";

describe("adaptManifestToMotionGraph", () => {
  it("maps the complete manifest graph to the hand-written canonical golden", () => {
    const graph = adaptManifestToMotionGraph(
      validateCompiledManifest(validManifest())
    );

    expect(graph.definition).toEqual({
      initialState: "a-a",
      states: [
        {
          id: "a-a",
          body: {
            unitId: "body-a",
            kind: "loop",
            frameCount: 4,
            ports: [
              { id: "default", entryFrame: 0, portalFrames: [0, 2] }
            ]
          },
          initialUnit: { unitId: "intro-a", frameCount: 2 }
        },
        {
          id: "a-b",
          body: {
            unitId: "body-b",
            kind: "finite",
            frameCount: 3,
            ports: [{ id: "default", entryFrame: 0, portalFrames: [2] }]
          }
        },
        {
          id: "a-c",
          body: {
            unitId: "body-c",
            kind: "held",
            frameCount: 1,
            ports: [{ id: "default", entryFrame: 0, portalFrames: [0] }]
          }
        }
      ],
      edges: [
        {
          id: "edge-ab",
          from: "a-a",
          to: "a-b",
          trigger: { type: "event", name: "go-b" },
          start: {
            type: "portal",
            sourcePort: "default",
            targetPort: "default",
            maxWaitFrames: 1
          },
          transition: {
            kind: "locked",
            unitId: "bridge-ab",
            frameCount: 2
          },
          continuity: "exact-authored"
        },
        {
          id: "edge-ac",
          from: "a-a",
          to: "a-c",
          trigger: { type: "event", name: "go-c" },
          start: { type: "cut", targetPort: "default", maxWaitFrames: 1 },
          continuity: "cut"
        },
        {
          id: "edge-ba",
          from: "a-b",
          to: "a-a",
          trigger: { type: "completion" },
          start: { type: "finish", targetPort: "default", maxWaitFrames: 2 },
          continuity: "exact-authored"
        },
        {
          id: "edge-bc",
          from: "a-b",
          to: "a-c",
          trigger: { type: "event", name: "go-c" },
          start: {
            type: "portal",
            sourcePort: "default",
            targetPort: "default",
            maxWaitFrames: 2
          },
          transition: {
            kind: "reversible",
            unitId: "rev-bc",
            frameCount: 6,
            direction: "forward"
          },
          continuity: "exact-authored"
        },
        {
          id: "edge-cb",
          from: "a-c",
          to: "a-b",
          trigger: { type: "event", name: "go-b" },
          start: {
            type: "portal",
            sourcePort: "default",
            targetPort: "default",
            maxWaitFrames: 0
          },
          transition: {
            kind: "reversible",
            unitId: "rev-bc",
            frameCount: 6,
            direction: "reverse",
            reverseOf: "edge-bc"
          },
          continuity: "exact-reverse"
        }
      ]
    });
  });

  it("returns a recursively immutable validated graph detached from the manifest", () => {
    const manifest = validateCompiledManifest(validManifest());
    const graph = adaptManifestToMotionGraph(manifest);

    expect(graph.definition.states).not.toBe(manifest.states);
    expect(Object.isFrozen(graph)).toBe(true);
    expect(Object.isFrozen(graph.definition)).toBe(true);
    expect(Object.isFrozen(graph.definition.states[0]?.body.ports[0]?.portalFrames)).toBe(true);
    expect(Object.isFrozen(graph.definition.edges[4]?.transition)).toBe(true);
  });

  it("wraps graph geometry and ambiguity failures as GRAPH_INVALID", () => {
    const manifest = structuredClone(validManifest()) as any;
    manifest.edges[0].start.maxWaitFrames = 0;
    const schemaValid = validateCompiledManifest(manifest);

    expect(() => adaptManifestToMotionGraph(schemaValid)).toThrowError(
      expect.objectContaining({ name: "FormatError", code: "GRAPH_INVALID" })
    );
  });

  it("wraps malformed trusted input without leaking built-in errors", () => {
    const malformed = {
      ...validManifest(),
      units: []
    } as unknown as CompiledManifest;

    try {
      adaptManifestToMotionGraph(malformed);
      throw new Error("expected graph adaptation to fail");
    } catch (error) {
      expect(error).toBeInstanceOf(FormatError);
      expect((error as FormatError).code).toBe("GRAPH_INVALID");
    }
  });
});
